internal/surfer/gcloud/utils/resource.go

```diff
@@ -0,0 +1,177 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/googleapis/librarian/internal/sidekick/api"
+	"github.com/iancoleman/strcase"
+)
+
+// GetPluralFromSegments infers the plural name of a resource from its structured path segments.
+// Per AIP-122, the plural is the literal segment before the final variable segment.
+// Example: `.../instances/{instance}` -> "instances".
+func GetPluralFromSegments(segments []api.PathSegment) string {
+	if len(segments) >= 2 {
+		lastSegment := segments[len(segments)-1]
+		if lastSegment.Variable != nil {
+			// The second to last segment should be the literal plural name
+			secondLastSegment := segments[len(segments)-2]
+			if secondLastSegment.Literal != nil {
+				return *secondLastSegment.Literal
+			}
+		}
+	}
+	return ""
+}
+
+// GetSingularFromSegments infers the singular name of a resource from its structured path segments.
+// The singular is the name of the final variable segment.
+// Example: `.../instances/{instance}` -> "instance".
+func GetSingularFromSegments(segments []api.PathSegment) string {
+	if len(segments) > 0 {
+		last := segments[len(segments)-1]
+		if last.Variable != nil && len(last.Variable.FieldPath) > 0 {
+			// Typically the variable name is the last component of the field path
+			// e.g. for `name` binding it might be implied? No, httprule parser populates FieldPath.
+			return last.Variable.FieldPath[len(last.Variable.FieldPath)-1]
+		}
+	}
+	return ""
+}
+
+// GetCollectionPathFromSegments constructs the base gcloud collection path from a
+// structured resource pattern, according to AIP-122 conventions.
+// It joins the literal collection identifiers with dots.
+// Example: `projects/{project}/locations/{location}/instances/{instance}` -> `projects.locations.instances`.
+func GetCollectionPathFromSegments(segments []api.PathSegment) string {
+	var collectionParts []string
+	for i := 0; i < len(segments)-1; i++ {
+		// A collection identifier is a literal segment followed by a variable segment.
+		if segments[i].Literal != nil && segments[i+1].Variable != nil {
+			collectionParts = append(collectionParts, *segments[i].Literal)
+		}
+	}
+	return strings.Join(collectionParts, ".")
+}
+
+// IsPrimaryResource determines if a field represents the primary resource of a method.
+func IsPrimaryResource(field *api.Field, method *api.Method) bool {
+	if method.InputType == nil {
+		return false
+	}
+	// For `Create` methods, the primary resource is identified by a field named
+	// in the format "{resource}_id" (e.g., "instance_id").
+	if strings.HasPrefix(method.Name, "Create") {
```

Comment:
We should have just one place that owns the logic of determining the type from the method name, if we decide HasPrefix is not enough, then we don't want to have to find them all.